c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -328,6 +328,35 @@ fn run_polonius<'tcx>(
     (facts, maps, output)
 }
 
+fn construct_adt_origins<'tcx>(
+    ltcx: &LTyCtxt<'tcx>,
+    adt_metadata: &AdtMetadataTable,
+    ty: &Ty,
+    amaps: &mut AtomMaps,
+) -> &'tcx [(OriginParam, Origin)] {
+    eprintln!("ty: {ty:?}");
+    let adt_def = ty.ty_adt_def().unwrap();
+
+    // create a concrete origin for each actual or hypothetical
+    // lifetime parameter in this ADT
+    let origins: Vec<_> = adt_metadata
+        .table
+        .get(&adt_def.did())
+        .map(|adt| {
+            adt.lifetime_params
+                .iter()
+                .map(|o| {
+                    let pairing = (*o, amaps.origin());
+                    eprintln!("pairing lifetime parameter with origin: {pairing:?}");
+                    pairing
+                })
+                .collect()
+        })
+        .unwrap_or_default();
+
+    ltcx.arena().alloc_from_iter(origins)
```

Comment:
```suggestion
    let default = Default::default();
    let origins = adt_metadata
        .table
        .get(&adt_def.did())
        .map_or(&default, |adt| &adt.lifetime_params)
        .iter()
        .map(|origin| (*origin, amaps.origin()))
        .inspect(|pairing| eprintln!("pairing lifetime parameter with origin: {pairing:?}"));
    ltcx.arena().alloc_from_iter(origins)
```

Don't need the `.collect::<Vec<_>>()`.